File: sqlbuilder/sqlbuilder/Binding.java
package com.clickability.cms.dataaccess.sqlbuilder;

import com.clickability.cms.type.DesignType;

import java.sql.Date;
import java.sql.Time;
import java.sql.Timestamp;
import java.sql.Types;

/**
 * For binding PreparedStatement '?' to actual values.
 * 
 * Add types as needed.
 * 
 */
public class Binding {

	public final BindingType type;
	public final Object value;
	
	// Add more if you need to.  You'll need an enum and a constructor.
	public enum BindingType {
		Array("setArray", Types.ARRAY, char[].class),
		Boolean("setBoolean", Types.TINYINT, boolean.class, Boolean.class),
		Byte("setByte", Types.TINYINT, byte.class, Byte.class),
		Char("setChar", Types.CHAR, char.class, Character.class),
		Short("setShort", Types.SMALLINT, short.class, Short.class),
		Int("setInt", Types.INTEGER, int.class, Integer.class),
		Long("setLong", Types.BIGINT, long.class, Long.class),
		Float("setFloat", Types.REAL, float.class, Float.class),
		Double("setDouble", Types.DOUBLE, double.class, Double.class),
		String("setString", Types.VARCHAR, String.class, DesignType.class),
		Date("setDate", Types.DATE, Date.class, java.util.Date.class),
		Time("setTime", Types.TIME, Time.class),
		TimeStamp("setTimestamp", Types.TIMESTAMP, Timestamp.class);
		
		public final String method;
		public final Class<?>[] classes;
		public final int sqlType;
		BindingType(String method, int type, Class<?>...classes) {
			this.method = method;
			this.classes = classes;
			this.sqlType = type;
		}
	}
	
	public Binding(Object v) {
		this.type = toBindingType(v.getClass());
		this.value = toSqlType(v);
	}
	
	// perform any supported conversions here, like
	// java.util.Date to java.sql.Timestamp
	public static Object toSqlType(Object value) {
		if (value.getClass().equals(java.util.Date.class)) {
			return new Timestamp(((java.util.Date)value).getTime());
		}

		// We support clickability common types too!
		else if (value.getClass().equals(DesignType.class)) {
			return ((DesignType)value).value();
		}
		return value;
	}
	
	// FIXME make sure value class is consistent.
	public static Object fromSqlType(Object value, Class<?> type) {
		if (value == null) return value;
		if (type.equals(DesignType.class)) {
			return DesignType.fromValue((String)value);
		}
		if (type.equals(java.util.Date.class)) {
			return new java.util.Date(((Timestamp)value).getTime());
		}
		return value;
	}
	
	public Binding(Class<?> type) {
		this.type = toBindingType(type);
		this.value = null;
	}
	
	private BindingType toBindingType(Class<?> type) {
		for (BindingType bt : BindingType.values()) {
			for (Class<?> clazz : bt.classes) {
				if (type.equals(clazz)) {
					return bt;
				}
			}
		}
		throw new DataaccessException(String.format("binding type %s not supported", type.getName()));
	}
	
}

File: queryBuilder/src/main/kotlin/queryBuilder/com/inunison/lib/queryBuilder/Binding.kt
package com.clickability.cms.dataaccess.sqlbuilder

import com.clickability.cms.type.DesignType

import java.sql.Date
import java.sql.Time
import java.sql.Timestamp
import java.sql.Types

/**
 * For binding PreparedStatement '?' to actual values.
 *
 * Add types as needed.
 *
 */
class Binding {

    val type: BindingType
    val value: Object?

    // Add more if you need to.  You'll need an enum and a constructor.
    enum class BindingType private constructor(val method: String, val sqlType: Int, vararg classes: Class<*>) {
        Array("setArray", Types.ARRAY, CharArray::class.java),
        Boolean("setBoolean", Types.TINYINT, Boolean::class.javaPrimitiveType, Boolean::class.java),
        Byte("setByte", Types.TINYINT, Byte::class.javaPrimitiveType, Byte::class.java),
        Char("setChar", Types.CHAR, Char::class.javaPrimitiveType, Character::class.java),
        Short("setShort", Types.SMALLINT, Short::class.javaPrimitiveType, Short::class.java),
        Int("setInt", Types.INTEGER, Int::class.javaPrimitiveType, Integer::class.java),
        Long("setLong", Types.BIGINT, Long::class.javaPrimitiveType, Long::class.java),
        Float("setFloat", Types.REAL, Float::class.javaPrimitiveType, Float::class.java),
        Double("setDouble", Types.DOUBLE, Double::class.javaPrimitiveType, Double::class.java),
        String("setString", Types.VARCHAR, String::class.java, DesignType::class.java),
        Date("setDate", Types.DATE, Date::class.java, java.util.Date::class.java),
        Time("setTime", Types.TIME, Time::class.java),
        TimeStamp("setTimestamp", Types.TIMESTAMP, Timestamp::class.java);

        val classes: Array<Class<*>>

        init {
            this.classes = classes
        }
    }

    constructor(v: Object) {
        this.type = toBindingType(v.getClass())
        this.value = toSqlType(v)
    }

    constructor(type: Class<*>) {
        this.type = toBindingType(type)
        this.value = null
    }

    private fun toBindingType(type: Class<*>): BindingType {
        for (bt in BindingType.values()) {
            for (clazz in bt.classes) {
                if (type.equals(clazz)) {
                    return bt
                }
            }
        }
        throw DataaccessException(String.format("binding type %s not supported", type.getName()))
    }

    companion object {

        // perform any supported conversions here, like
        // java.util.Date to java.sql.Timestamp
        fun toSqlType(value: Object): Object {
            if (value.getClass().equals(java.util.Date::class.java)) {
                return Timestamp((value as java.util.Date).getTime())
            } else if (value.getClass().equals(DesignType::class.java)) {
                return (value as DesignType).value()
            }
            return value
        }

        // FIXME make sure value class is consistent.
        fun fromSqlType(value: Object?, type: Class<*>): Object? {
            if (value == null) return value
            if (type.equals(DesignType::class.java)) {
                return DesignType.fromValue(value as String?)
            }
            return if (type.equals(java.util.Date::class.java)) {
                java.util.Date((value as Timestamp).getTime())
            } else value
        }
    }

}
